### PR TITLE
Sh combobox improvements

### PR DESCRIPTION
**Pull request by cpinter**
1. Introduce None item in subject hierarchy
    - If enabled, an "extra item" (similarly to the same concept in qMRMLNodeComboBox) is added as the top item under the scene. This item has the name defined in the noneItemText property, and is selectable but no drag&drop is enabled. When selecting it, the current item will have the ID 0 (= INVALID_ITEM_ID)
    - This enables having empty selections in comboboxes after having made a valid selection

    image

2. Add flag in subject hierarchy combobox for showing parents in title
    - Sometimes it is undesired to see the name of each parent after selecting an item in the combobox (it looks like "[ParentName] / [ParentName] / [SelectedItemName]"). If disabled, only the name of the selected item is shown ("[SelectedItemName]")
    - On by default not to change the current behavior

3. If an include attribute node filter was applied, hierarchy (patient/study/folder) items that were left without shown children were still visible, while they were supposed to only be shown if any child was shown

4. If an include attribute node filter was applied, hierarchy items were still selectable, but due to the filter only the items accepted based on the filter are supposed to be interacted with

### CODE REVIEW

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx
Comment (lassoan) [inline Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx:475]: It would be better to have the same API as qMRMLNodeComboBox. So, the method name should be `noneEnabled`.

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx
Comment (lassoan) [inline Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx:485]: `setNoneEnabled`

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx
Comment (lassoan) [inline Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx:495]: It would be better to mirror qMRMLNodeComboBox API -> rename to `noneDisplay`.

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx
Comment (lassoan) [inline Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx:505]: It would be better to mirror qMRMLNodeComboBox API -> rename to `setNoneDisplay`.

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.h
Comment (lassoan) [inline Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.h:69]: It is not really a title. In a Qt combo-box it is called the "lineEdit", which might not the most intuitive name either, but at least not we invented the name just use an existing name.

**Review by lassoan — APPROVED** [review]
It looks good, I've added just a few small comments. It would be nice if you could add at least some minimal test, which would also server as an example.

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx
Comment (cpinter) [reply Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx:495]: First I followed these function names but then I thought that although the `noneEnabled` and `noneDisplay` are unambiguous in a combobox, they are (especially `noneDisplay`) not very expressive of what they does in a subject hierarchy model or tree view. Because I had to introduce the same properties in these two other classes as well. I'll wait for your answer before changing these.

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.h
Comment (cpinter) [reply Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.h:69]: I just tried to follow the convention I found. See `updateComboBoxTitleAndIcon` function. At least it is definitely not a non-existent name.

**Comment by cpinter** [comment]
Thanks for the review @lassoan ! I added a test, but had some comments about your renaming requests so I'll wait until we settle on those before I push.

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx
Comment (lassoan) [reply Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx:495]: I don't particularly like the existing names but it is important to make this new combobox a drop-in replacement of vtkMRMLNodeComboBox. If you want, you can add alternative names (different methods all calling the same underlying method).

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.h
Comment (lassoan) [reply Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.h:69]: `updateComboBoxTitleAndIcon` is not in the public API, so it was not important there to have a meaningful or consistent name. I don't have much better suggestions, so if you don't like any of the ones that I listed below any better then I'm OK with the current name, too.

showCurrentItemParents
showCurrentItemParentsInLineEdit
showParentsInLineEdit

**Comment by lassoan** [comment]
Now that you are modifying this code (add a top-level item), would you be able to add back the top-level "Scene" item? It would be really useful to be able to access scene-level features (that now requires clicking in a "white" area in the SH tree that even people who know what to look for have often trouble finding).

**Comment by cpinter** [comment]
> add back the top-level "Scene" item

Do you mean by default to the SH tree in the Data module? Yes the white area issue is quite annoying. I think we removed it back then to save valuable horizontal space, but I don't see a problem re-adding it.

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx
Comment (cpinter) [reply Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx:495]: That's a good point about trivial switch from qMRMLNodeComboBox. I changed these properties and functions back.

File: Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.h
Comment (cpinter) [reply Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.h:69]: I renamed this to `showCurrentItemParents`. I think LineEdit is even worse than Title because it cannot be edited in this case.

**Comment by cpinter** [comment]
I changed everything we discussed. Once we agree on the scene item issue I'll add a commit in this PR for it.

**Comment by lassoan** [comment]
> > add back the top-level "Scene" item
> 
> I probably wouldn't change the default not to modify all the SH trees used in all the modules/extensions/etc.
> 
> However, the white area issue is quite annoying. I think we removed it back then to save valuable horizontal space, but I don't see a problem re-adding it to the Data module (Maybe the DICOM module as well? Maybe not because it is mainly intended for seeing the scene content)

You are right, it makes sense to keep the current behavior by default (to not impact existing modules that might not expect the change), but we can enable it explicitly when we know we have enough space.

**Comment by cpinter** [comment]
I re-added the scene item only in the Data module. I reduced the margins from 9/9 to 4/4 within the tab widget tabs (as already the case in other modules where space is important). It looks like this:
image

@lassoan Please approve if it looks good to you. Thanks!

**Comment by cpinter** [comment]
(Note: I think back when we removed the scene item we had the transform column show the transform names instead of the icons so horizontal space looked very restricted)

**Review by lassoan — APPROVED** [review]
Looks good, thank you very much.

**Pull request merged**
